File: app/src/main/java/tech/pauly/findapet/discover/FilterActivity.java
package tech.pauly.findapet.discover;

import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearSmoothScroller;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import javax.inject.Inject;

import dagger.android.AndroidInjection;
import io.reactivex.disposables.CompositeDisposable;
import tech.pauly.findapet.R;
import tech.pauly.findapet.databinding.ActivityFilterBinding;
import tech.pauly.findapet.shared.BaseActivity;
import tech.pauly.findapet.shared.events.ViewEventBus;

public class FilterActivity extends BaseActivity {

    @Inject
    FilterViewModel viewModel;

    @Inject
    ViewEventBus eventBus;

    private ActivityFilterBinding binding;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_filter);
        getLifecycle().addObserver(viewModel);
        binding.setViewModel(viewModel);

        setSupportActionBar(findViewById(R.id.toolbar));
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_close);
        }
    }

    @Override
    public void onResume() {
        super.onResume();

        subscribeOnLifecycle(viewModel.getScrollToViewSubject()
                                      .subscribe(this::scrollToBreedSearch, Throwable::printStackTrace));
    }

    private void scrollToBreedSearch(Boolean b) {
        RecyclerView.SmoothScroller smoothScroller = new LinearSmoothScroller(this) {
            @Override protected int getVerticalSnapPreference() {
                return LinearSmoothScroller.SNAP_TO_START;
            }
        };
        smoothScroller.setTargetPosition(1);
        binding.filterRecyclerView.getLayoutManager().startSmoothScroll(smoothScroller);
    }

    @Nullable
    @Override
    protected CompositeDisposable registerViewEvents() {
        CompositeDisposable viewEvents = new CompositeDisposable();

        viewEvents.add(eventBus.activity(FilterViewModel.class).subscribe(this::activityEvent));

        return viewEvents;
    }
}

File: app/src/main/java/tech/pauly/findapet/discover/FilterActivity.kt
package tech.pauly.findapet.discover

import android.content.Context
import android.os.Bundle
import android.view.View
import android.view.animation.AnimationUtils
import android.view.inputmethod.InputMethodManager
import androidx.core.content.ContextCompat.getSystemService
import androidx.databinding.DataBindingUtil
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.plusAssign
import kotlinx.android.synthetic.main.activity_filter.*
import tech.pauly.findapet.R
import tech.pauly.findapet.databinding.ActivityFilterBinding
import tech.pauly.findapet.dependencyinjection.PetApplication
import tech.pauly.findapet.shared.BaseActivity
import tech.pauly.findapet.shared.events.ViewEventBus
import tech.pauly.findapet.utils.addAnimationListener
import tech.pauly.findapet.utils.runAnimation
import javax.inject.Inject

class FilterActivity : BaseActivity() {

    @Inject
    lateinit var viewModel: FilterViewModel

    @Inject
    lateinit var eventBus: ViewEventBus

    private lateinit var breedViewModel: BreedViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        PetApplication.component.inject(this)
        super.onCreate(savedInstanceState)
        val binding = DataBindingUtil.setContentView<ActivityFilterBinding>(this, R.layout.activity_filter)
        addViewModelLifecycleObserver(viewModel)
        binding.viewModel = viewModel
        breedViewModel = viewModel.breedViewModel.also {
            addViewModelLifecycleObserver(it)
            binding.breedViewModel = it
        }

        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setHomeAsUpIndicator(R.drawable.ic_close)
        }
    }

    override fun onResume() {
        super.onResume()
        handleSearch()
    }

    private fun handleSearch() {
        breedViewModel.openBreedSubject.quickSubscribe { open ->
            if (open) showBreed() else hideBreed()
        }
    }

    override val viewEvents: CompositeDisposable?
        get() = CompositeDisposable().also {
            it += eventBus.activity(FilterViewModel::class).subscribe(this::activityEvent)
        }

    private fun showBreed() {
        breed_layout.visibility = View.VISIBLE
        val inputManager = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        breed_layout.runAnimation { _ ->
            AnimationUtils.loadAnimation(this, R.anim.scroll_bottom_to_center)
                    .addAnimationListener(
                            animationEnd = {
                                breed_search_edit_text.requestFocus()
                                inputManager.toggleSoftInput(InputMethodManager.SHOW_FORCED, InputMethodManager.HIDE_IMPLICIT_ONLY)
                            })
        }
    }

    private fun hideBreed() {
        val inputManager = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        breed_layout.runAnimation { layout ->
            AnimationUtils.loadAnimation(this, R.anim.scroll_center_to_bottom)
                    .addAnimationListener(
                            animationStart = {
                                inputManager.hideSoftInputFromWindow(breed_search_edit_text.windowToken, 0)
                                breed_search_edit_text.clearFocus()
                            },
                            animationEnd = {
                                layout.visibility = View.GONE
                            })
        }
    }
}
